datalad_next/iter_collections/annexworktree.py

```diff
@@ -0,0 +1,123 @@
+    Any,
+    Generator,
+)
+
+from more_itertools import intersperse
+
+from .gitworktree import (
+    GitWorktreeItem,
+    GitWorktreeFileSystemItem,
+    iter_gitworktree
+)
+from datalad_next.itertools import (
+    itemize,
+    join_with_list,
+    load_json,
+    route_in,
+    route_out,
+    dont_process,
+)
+from datalad_next.runners import iter_subproc
+
+
+lgr = logging.getLogger('datalad.ext.next.iter_collections.annexworktree')
+
+
+linesep_bytes = linesep.encode()
+
+
+@dataclass
+class AnnexWorktreeItem(GitWorktreeItem):
+    annexkey: str | None = None
+    annexsize: int | None = None
+    annexobjpath: PurePath | None = None
+
+
+@dataclass
+class AnnexWorktreeFileSystemItem(GitWorktreeFileSystemItem):
+    annexkey: str | None = None
+    annexsize: int | None = None
+
+
+def iter_annexworktree(
+        path: Path,
+        *,
+        untracked: str | None = 'all',
+        link_target: bool = False,
+        fp: bool = False,
+) -> Generator[AnnexWorktreeItem | AnnexWorktreeFileSystemItem, None, None]:
+
+    glsf = iter_gitworktree(
+        path,
+        untracked=untracked,
+        link_target=link_target,
+        fp=fp
+    )
+
+    git_fileinfo_store: list[Any] = list()
+    key_store: list[Any] = list()
+
+    with \
+            iter_subproc(
+                # we get the annex key for any filename (or empty if not annexed)
+                ['git', '-C', str(path), 'annex', 'find', '--anything', '--format=\${key}\n', '--batch'],
+                # intersperse items with newlines to trigger a batch run
+                # this avoids string operations to append newlines to items
+                input=intersperse(
+                    b'\n',
+                    # store all output of the git ls-find in the gitfileinfo
+                    # store
+                    route_out(
+                        glsf,
+                        git_fileinfo_store,
+                        lambda data: (str(data.name).encode(), [data])
+                    )
+                ),
+            ) as gaf, \
+            iter_subproc(
+                # get the key properties JSON-lines style
+                ['git', '-C', str(path), 'annex', 'examinekey', '--json', '--batch'],
+                # process all non-empty keys and store them in the key store,
+                # skip processing of empty keys and store an ignored value in
+                # the key store
+                input=route_out(
+                    itemize(gaf, sep=linesep_bytes, keep_ends=True),
+                    key_store,
+                    lambda data: (dont_process, [None])
+                                 if data == linesep_bytes
+                                 else (data, [data])
+                )
+            ) as gek:
+
+        for item in route_in(
+                route_in(
+                    load_json(itemize(gek, sep=linesep_bytes)),
+                    key_store,
+                    join_with_list,
+                ),
```

Comment:
What is the purpose of this `route_in`? It seems to join the keyname with the output record of `examinekey` for that keyname. However, the keyname is already present in that record.

Maybe a comment can clarify the purpose.

Reply:
The purpose is twofold:

1. inform the consumer whether a key was present for an item of `git_fileinfo_store` or not. (If it was present, processing through `examinekey` was peformed and the key record will be returned.  If it was not present, `examinekey` was not performed and `None` will be returned.)
2. "upscale" the results of gek (# annexed files) to the cardinality of `git_fileinfo_store` (# git files)

Reply:
Comment updated in 1e1d3e57fab42ebd1b9b20c1e6e5d91ea69473ee